ADR91 / AP18 content (#90)
bf2fc6cc711aee1a0c2a/architecture: _ap/18/index.adoc

Additionally, the fluent code style of the generated clients provides a better experience for consumers, resulting in higher levels of developer joy.

==== Maintenance

Experience shows that issues are answered within hours, PRs are continuously accepted and merged, and guidance is provided when needed.

Transitive dependencies are kept up to date in a timely manner.

Improvements and bugs are continuously integrated into both the generator and the auxiliary libraries' repositories.

==== Swappable implementation libraries

The design of Kiota allows consumers of the code generator to provide their own https://microsoft.github.io/kiota/extending/corelibrary.html[“opinionated” implementations] of core generated client functionality.
This is done via a set of interfaces to core functionality that all generated clients leverage (e.g. a Java HTTP client interface).  Default implementations of these interfaces are provided, but alternatives can easily be created and used.

> Specifically for the Red Hat use case, we are free to roll out our own implementations of the core libraries to better support Red Hat products and strategy (e.g. Quarkus specific implementations of certain functionality).

Reviewer: It would be good to spell out the implementations we're expecting to write. HTTP client is one, but what about authentication, for example?

Author: Good question, regarding implementations, for Java I'd expect:
- a Vert-X HTTP client implementation (as opposed to the default OkHttp)
- a JSON serialization based on Jackson (as opposed to the default Gson)

Do we have strong opinions on the other languages @tombentley ?

@riccardo-forina can you check the TS libraries? https://github.com/microsoft/kiota-typescript
@dimakis can you have a look over Python? https://github.com/microsoft?q=kiota+python&type=all&language=&sort=
@lburgazzoli can I ask for your support in Go? https://github.com/microsoft?q=kiota+go&type=all&language=&sort=

🙏  thanks in advance!

Regarding Authentication I think that we should support only OFFLINE_TOKEN, this means rolling out a custom `TokenProvider` for each language (it's not mandatory but highly conveninet). this is in Java for example:
https://github.com/andreaTP/kiota-utils#redhat-sso-auth

Author: @tombentley addressed here: https://github.com/bf2fc6cc711aee1a0c2a/architecture/pull/90/commits/c9f2cee1fc624782620d4ca59498b5bf64c50ca1 let me know if you prefer additional detail.

Creating alternatives for the provided default libraries is doable in a very reasonable amount of time.
https://github.com/andreaTP/kiota-utils/blob/1a7dcecd92ce8794fe479e9c37c71d5733a39ff6/http/jdk-http/src/main/java/io/apicurio/kiota/http/JdkHttpRequestAdapter.java#L1[Here] you can find an alternative implementation of the Java HttpClient interface that is based on the client embedded in JDK 11+ (vs. the out-of-the-box implementation that is based on OkHttp).

In the long term, we expect that most of the bug fixes are going to happen in those shared libraries, and being able to maintain and release them at our own pace is a key element for success.

==== Compatibility matrix

Kiota https://github.com/microsoft/kiota/issues/2020[compatibility matrix] covers the current Red Hat use-cases, specifically:

* Java
* Go
* Python

> * TypeScript

Reviewer 2: ```suggestion
* JavaScript
```

typo?

Reviewer 2: line just below is TypeScript, so TypeScript was listed twice

Author: thanks for the suggestion!
This is actually deliberate, even the current "js-sdk" is producing TS:
https://github.com/redhat-developer/app-services-sdk-js

=== Threat Model

image::threat-model.png[]

* Given that Kiota is going to produce source code and runs on the client side there are no specific threats for RH SaaS
* A Code Generator could produce malicious or corrupted generated code.
* Clients with large numbers of dependencies could be vulnerable if not refreshed
* A swapped implementation may not be well supported

=== Alternatives Considered / Rejected

==== Keep going with the current generator

The DevExp team has identified as a major piece of maintenance the upgrade of the currently outdated https://github.com/OpenAPITools/openapi-generator[openapi-generator].
The current SDKs are full of quirks and edge cases in most cases hacked around to end up with a working implementation.
Having to keep the https://docs.google.com/spreadsheets/d/1fwmRmapDdFRU-B7qFBUuXyHMDjzXtRMY-jhbtxDVtyU/edit#gid=0[hacks and workarounds] is a major blocker to updating toward newer versions of the generator.

The monolithic approach to a generation used by openapi-generator also offers challenges in a modular approach to bug fixes and improvements.

==== IBM

> We have had extensive discussions with IBM regarding their approach to https://github.com/IBM/ibm-cloud-sdk-common[client SDK] generation and we very much like their governance model over APIs produced in the company.

Reviewer 3: I am missing why this was ruled out?

Author: The section on IBM was based on confidential information.
I stripped it down to a minimum that can be safely shared.

Which are the options to keep part of this document private?
Should I make an "Addendum" in the form of a Google Doc linked here?
@tombentley do you have any suggestion here?

Reviewer: GDoc seems like a good idea to me @andreaTP

Author: done: https://github.com/bf2fc6cc711aee1a0c2a/architecture/pull/90/commits/6bf32c97b56f4c73c3b363481f23ef38b8e854b6

* CLI tools
* UI (e.g. console.dot)
* Provisioning Tools (Ansible, Terraform)

Most of our toolchains are (in some ways) using `openapi-generator`, developed and maintained by the "OpenAPI Tools" organization (which is NOT affiliated with the OpenAPI initiative).

Unfortunately, OpenAPI Generator is not a well maintained OSS project:
image::openapi-generator-gh.png[]

Even our own attempts to fix known bugs have been ineffective:
https://github.com/OpenAPITools/openapi-generator/pull/13616

It's also pretty common to have breaking changes in the generator tooling, which results in breaking changes in the generated code.
When this happens it can be hard to detect and often happens without a clear migration path.

For these reasons we are often https://github.com/redhat-developer/app-services-sdk-go/blob/13d1cad9a937c456a9623ad188c44790e3370d83/openapitools.json#L5[stuck on old versions], preventing (in some cases) the opportunity to fix CVEs and issues by keeping transitive dependencies and used frameworks updated.

Following are a variety of critical/important issues we currently have due to our OpenAPI Generator based approach to API Client generation.

=== Security

> 

Reviewer 4: Discuss mitigation on security issues coming from. Probably iso risk between old and new approach but make it clear of the security risk paths.

Reviewer 4: Ah it's in the Thread Model section

=== Community

We want our community of customers to be able to use best-in-class tools to perform their day-to-day work. OpenAPI generation is an important piece of tooling and we should be able to provide a much better OSS experience than the one we are right now.

== Goals

* Identify alternative tooling that can be used in place of the current OpenAPI generator
* Validate and improve the integration patterns and the user experience
* Use up-to-date tooling and libraries for our customer-facing API clients

== Non-goals

* Dictate a technical decision for internal usage or for projects that are successfully rolling their own SDKs and clients (e.g. by manually curating them)

== Current Architecture

Inside Red Hat we are mostly relying (in various ways) on OpenAPI Generator to generate clients, we also maintain a https://github.com/quarkiverse/quarkus-openapi-generator[Quarkus Extension] based on the same generator.

== Proposed Architecture

> We propose to switch to alternative OSS tooling provided by Microsoft called https://github.com/microsoft/kiota[Kiota].

Reviewer 4: ```suggestion
We propose to switch to an alternative OSS tooling provided by Microsoft called https://github.com/microsoft/kiota[Kiota].
```

=== Advantages:

* Community
* Better design/architecture
* Better developer experience
* Maintenance
* Swappable implementation libraries
* Compatibility matrix

=== Disadvantages:

* Not GA (yet)
* External Owners (Microsoft)

> * Core Generator codebase is .NET

Reviewer 5: It is not a key aspect for this ADR but does it imply that this library cannot be integrated with Quarkus to replace the existing Quarkus OpenAPI extension or in general any possible Quarkus similar use cases?

Author: Let me try to provide more context here:

- `this library cannot be integrated with Quarkus` -> Kiota ships as a standalone CLI + Java dependencies and will work with Quarkus (possibly it will need a little bit of tweaking for `native-image`, but it should be it). We are already publishing a Maven plugin to easily integrate with Java workflows: https://github.com/redhat-developer/kiota-java-extra#maven-plugin
- `replace the existing Quarkus OpenAPI extension` -> replacing the current extension is out of the scope of this AP and out of the scope of Kiota. The Quarkus extension is vertical and highly integrated with an opinionated language and framework, while Kiota privileges a more consistent approach across different languages and decouples over the used frameworks. I plan to agree on an action plan with Ricardo to see if it makes sense to support both approaches, but no strong commitment here.
- `or in general any possible Quarkus similar use cases?` -> can you, please, elaborate if there is any aspect that I might have missed with the previous answers?

Reviewer 5: Thanks for the clarification, you covered all my questions :)

The design of Kiota allows consumers of the code generator to provide their own https://microsoft.github.io/kiota/extending/corelibrary.html[“opinionated” implementations] of core generated client functionality.
This is done via a set of interfaces to core functionality that all generated clients leverage (e.g. a Java HTTP client interface).  Default implementations of these interfaces are provided, but alternatives can easily be created and used.

Specifically for the Red Hat use case, we are free to roll out our own implementations of the core libraries to better support Red Hat products and strategy (e.g. Quarkus specific implementations of certain functionality).

Creating alternatives for the provided default libraries is doable in a very reasonable amount of time.
https://github.com/andreaTP/kiota-utils/blob/1a7dcecd92ce8794fe479e9c37c71d5733a39ff6/http/jdk-http/src/main/java/io/apicurio/kiota/http/JdkHttpRequestAdapter.java#L1[Here] you can find an alternative implementation of the Java HttpClient interface that is based on the client embedded in JDK 11+ (vs. the out-of-the-box implementation that is based on OkHttp).
We expect to provide:

> - one `RHTokenProvider` that will work automatically with [offline tokens](https://access.redhat.com/articles/3626371) and SSO like [we did for Java](https://github.com/andreaTP/kiota-utils/blob/93083650e0374f511abefd099033f05276927deb/rh-auth/src/main/java/com/github/andreatp/kiota/auth/RHAccessTokenProvider.java#L24) for each supported language.

Reviewer 5: In the preview these links seems to be malformed (probably you are using markdown syntax instead of adoc)
```suggestion
- one `RHTokenProvider` that will work automatically with https://access.redhat.com/articles/3626371[offline tokens] and SSO like https://github.com/andreaTP/kiota-utils/blob/93083650e0374f511abefd099033f05276927deb/rh-auth/src/main/java/com/github/andreatp/kiota/auth/RHAccessTokenProvider.java#L24[we did for Java] for each supported language.
```

Author: thanks!

==== Legal Requirements (Terms Acceptance and Export Control)

Kiota and the relevant component are released under the MIT license.

=== Additional Reading

Kiota design: https://microsoft.github.io/kiota/extending/
Some libraries are already used [in Red Hat projects](https://github.com/microsoft/kiota-abstractions-go/network/dependents).

=== Threat Model

image::threat-model.png[]

* Given that Kiota is going to produce source code and runs on the client side there are no specific threats for RH SaaS
* A Code Generator could produce malicious or corrupted generated code.
* Clients with large numbers of dependencies could be vulnerable if not refreshed
* A swapped implementation may not be well supported

=== Alternatives Considered / Rejected

==== Keep going with the current generator

The DevExp team has identified as a major piece of maintenance the upgrade of the currently outdated https://github.com/OpenAPITools/openapi-generator[openapi-generator].
The current SDKs are full of quirks and edge cases in most cases hacked around to end up with a working implementation.

> Having to keep the https://docs.google.com/spreadsheets/d/1fwmRmapDdFRU-B7qFBUuXyHMDjzXtRMY-jhbtxDVtyU/edit#gid=0[hacks and workarounds] is a major blocker to updating toward newer versions of the generator.

Reviewer 5: This link is not publicly accessible, I don't know if it makes sense to link it here

Reviewer 5: Is the content of this link private? If so, fine for me to follow the same approach :)

Author: ouch, my bad, I haven't noticed this is actually a different link, I think we can safely drop it 👍

Author: removed

title: "OpenAPI clients code generation - Kiota"
status: "Draft"
authors:
- "Andrea Peruffo"
tags: []
---

## Intent

> Define a consistent technology to be used for client code generation from OpenAPI specifications.

Reviewer 4: ```suggestion
Describe a path to generate client code from OpenAPI specification for our service APIs.
Lower the burden of code generation from OpenAPI specifications
```

## Motivation

> * Have a uniform and good quality SDK layer to be used for APIs access

Reviewer 4: * offer a client or SDK for Services APIs that CLI, UI and customers can rely on

Author: The fact that I left this one out was actually intentional ... I can get it back, but my intention was to:

> leave the door open to (eventually)provide only "tooling" and not full-fledged SDKs

I understand this might sound a little excessive at this stage but, at least, I would like to keep this door open.
Thoughts?

Reviewer 4: When you leave the door open, you assume a lot of imagination from a casual reader looking to know if his problem is solved. And that reader ahs to read a lot of things so won't meditate on it. I would go for explicit, you can add tooling as an extra option but the main use case being the sentence you left out.

Author: ok, fair 👍

Author: added

* Be able to maintain the generated code and the necessary libraries
* Ability to fix bugs and eventually improve the generator codebase

## Applicability

> This pattern applies to any service using a client/SDK to access REST API endpoints.

Reviewer 4: ```suggestion
This pattern applies to any service providing a client/SDK to access REST API endpoints.
```

Author: this is actually not correct, Kiota is going to be used on the "receiver side", and possibly even without the need of curated SDKs

Reviewer 4: So the intent is not to offer SDKs (generated). But to use Kiota direcly in the CLI and UI build system?

Author: integrated with `using or providing`

Exceptions are:
 - a service is already successfully using an alternative generator
 - a service is using the current `quarkus-openapi-extension` (the current plan is to move the extension to use Kiota as well)

## Description

We propose to use OSS tooling provided by Microsoft called https://github.com/microsoft/kiota[Kiota] for client and SDKs generations.

To apply Kiota in a project follow the https://microsoft.github.io/kiota/get-started/[Official "Get Started" guides].

## Participants

> The DevExp team is currently providing SDKs for all of the MAS services, all of them will be converted to use Kiota and serve as an example that can be applied more broadly.

Reviewer 4: ```suggestion
The DevExp team is currently providing SDKs for all of the bf2 services, all of them will be converted to use Kiota and serve as an example that can be applied more broadly.
```
